src/wallet/walletdb.cpp

```diff
@@ -971,6 +1087,106 @@ DBErrors WalletBatch::LoadWallet(CWallet* pwallet)
         }
     }
 
+    // Upgrade to having a wallet hd key
+    // Find the keys which are used in single key internal and external descriptors with
+    // the pre-taproot output types
+    if (pwallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS) &&
+        !pwallet->IsWalletFlagSet(WALLET_FLAG_GLOBAL_HD_KEY)
+        ) {
+        std::map<CExtPubKey, std::pair<std::map<std::string, int>, uint64_t>> descs_keys;
+        std::map<std::string, int> tmpl = {{"pkh(", 0}, {"sh(wpkh(", 0}, {"wpkh(", 0}};
+
+        // Find root xpubs used in pkh(), sh(wpkh()), and wpkh() descriptors
+        for (const auto& spkm : pwallet->GetAllScriptPubKeyMans()) {
+            const DescriptorScriptPubKeyMan* desc_spkm = dynamic_cast<DescriptorScriptPubKeyMan*>(spkm);
+            assert(desc_spkm);
+            LOCK(desc_spkm->cs_desc_man);
+            WalletDescriptor w_desc = desc_spkm->GetWalletDescriptor();
+
+            if (!w_desc.descriptor->IsRange()) {
+                continue;
+            }
+
+            std::set<CPubKey> desc_pubkeys;
+            std::set<CExtPubKey> desc_xpubs;
+            w_desc.descriptor->GetPubkeys(desc_pubkeys, desc_xpubs);
+            if (desc_xpubs.size() != 1) {
+                continue;
+            }
+            const CExtPubKey& xpub = *desc_xpubs.begin();
+
+            if (descs_keys.count(xpub) == 0) {
+                descs_keys.emplace(xpub, std::make_pair(tmpl, 0));
+            }
+
+            const std::string desc = w_desc.descriptor->ToString();
+            if (desc.find("pkh(") == 0) {
+                descs_keys[xpub].first["pkh("]++;
+            } else if (desc.find("sh(wpkh(") == 0) {
+                descs_keys[xpub].first["sh(wpkh("]++;
+            } else if (desc.find("wpkh(") == 0) {
+                descs_keys[xpub].first["wpkh("]++;
+            } else {
+                continue;
+            }
+            if (w_desc.creation_time > descs_keys[xpub].second) {
+                descs_keys[xpub].second = w_desc.creation_time;
+            }
+        }
+
+        // Find candidate active xpubs
+        // These are the ones that are used in 2 pkh(), 2 sh(wpkh()), and 2 wpkh() descriptors
+        uint64_t best_time = 0;
+        CExtPubKey best_xpub;
+        std::map<CKeyID, CExtPubKey> xpubs;
+        for (const auto& [xpub, info] : descs_keys) {
+            const auto& [dtypes, desc_time] = info;
+            bool ok = true;
+            for (const auto& [dtype, count] : dtypes) {
+                if (count != 2) {
+                    ok = false;
+                    break;
+                }
+            }
```

Comment:
3e53cc298ca1037ad2f52c65934a5e71b76f82bd
Could be reduced to:

```suggestion
            bool ok = std::all_of(dtypes.begin(), dtypes.end(), [](const auto& p) { return p.second == 2; });
```

Reply:
Done